File: src/claw_mem/dreaming/rem.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Dreaming Engine — REM Phase (Pattern Extractor | v4.12.0)

Builds triplets from scored candidates, runs SkillExtractor,
and groups results into topic summaries by tag.
"""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .deep import ScoredCandidate


@dataclass
class REMResult:
    """Result of the REM pattern extraction phase.

    Attributes:
        triplets: List of extracted (s, p, o) dicts.
        skills: List of extracted Skill objects.
        topic_summaries: Dict mapping topic tag to summary string.
        extracted_count: Total number of triplets extracted.
        skills_count: Total number of skills extracted.
    """

    triplets: List[Dict[str, str]] = field(default_factory=list)
    skills: List[Any] = field(default_factory=list)
    topic_summaries: Dict[str, str] = field(default_factory=dict)
    extracted_count: int = 0
    skills_count: int = 0

    def to_dict(self) -> Dict[str, Any]:
        return {
            "extracted_count": self.extracted_count,
            "skills_count": self.skills_count,
            "topic_summaries": self.topic_summaries,
            "triplets": self.triplets,
        }


class PatternExtractor:
    """Extract patterns from scored dreaming candidates.

    Reuses SkillExtractor from claw_mem.extraction.skill_extractor for
    skill pattern abstraction. Groups candidates by tags and generates
    topic summaries.
    """

    def __init__(self, llm_provider: Any = None):
        self._llm = llm_provider

    def extract(self, candidates: List[ScoredCandidate]) -> REMResult:
        """Run pattern extraction on scored candidates.

        Args:
            candidates: Filtered ScoredCandidate list from deep phase.

        Returns:
            REMResult with triplets, skills, and topic summaries.
        """
        result = REMResult()

        # 1. Build triplets from candidate content
        triplets = self._build_triplets(candidates)
        result.triplets = triplets
        result.extracted_count = len(triplets)

        # 2. Run SkillExtractor
        skills = self._extract_skills(triplets)
        result.skills = skills
        result.skills_count = len(skills)

        # 3. Build topic summaries by tag
        result.topic_summaries = self._build_topic_summaries(candidates)

        return result

    # ── triplet construction ───────────────────────────────────────

    def _build_triplets(self, candidates: List[ScoredCandidate]) -> List[Dict[str, str]]:
        """Build simple (subject, predicate, object) triplets from candidates.

        Uses a simple heuristic: the first noun phrase is the subject,
        the middle portion implies the predicate, and remaining content
        becomes the object.
        """
        triplets: List[Dict[str, str]] = []
        for c in candidates:
            content = c.signal.content
            if not content:
                continue
            # Simple split heuristic
            parts = content.split(":", 1) if ":" in content else content.split("：", 1)
            if len(parts) == 2:
                subject = parts[0].strip()[:50]
                obj = parts[1].strip()[:100]
            else:
                words = content.split()
                if len(words) >= 3:
                    subject = " ".join(words[:2])[:50]
                    obj = " ".join(words[2:])[:100]
                else:
                    subject = content[:50]
                    obj = ""

            triplets.append({
                "s": subject,
                "p": "relates_to",
                "o": obj,
            })

        return triplets

    def _extract_skills(self, triplets: List[Dict[str, str]]) -> List[Any]:
        """Run SkillExtractor on the constructed triplets.

        Converts simple dict triplets to Triplet objects expected by
        SkillExtractor.
        """
        if not triplets:
            return []

        try:
            from claw_mem.extraction.openie_extractor import Triplet
            from claw_mem.extraction.skill_extractor import SkillExtractor

            # Convert dict triplets to Triplet objects
            triplet_objs = [
                Triplet(subject=t["s"], predicate=t["p"], object=t["o"], confidence=0.7)
                for t in triplets
            ]

            extractor = SkillExtractor(llm_provider=self._llm, mode="rule")
            return extractor.extract(triplet_objs)
        except Exception:
            return []

    # ── topic summaries ────────────────────────────────────────────

    def _build_topic_summaries(self, candidates: List[ScoredCandidate]) -> Dict[str, str]:
        """Group candidates by tags and build concise topic summaries."""
        topic_groups: Dict[str, List[str]] = defaultdict(list)

        for c in candidates:
            tags = c.signal.tags if c.signal.tags else ["general"]
            for tag in tags:
                topic_groups[tag].append(c.signal.content)

        summaries: Dict[str, str] = {}
        for tag, contents in topic_groups.items():
            if len(contents) == 1:
                summaries[tag] = contents[0][:100]
            else:
                # Concatenate first few items
                preview = " | ".join(c[:60] for c in contents[:5])
                suffix = f" ... (+{len(contents) - 5} more)" if len(contents) > 5 else ""
                summaries[tag] = f"[{len(contents)} items] {preview}{suffix}"

        return summaries

File: src/dreaming/rem.ts
// Copyright 2026 Peter Cheng
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Dreaming Engine -- REM Phase (Pattern Extractor | v4.12.0)
 *
 * Builds triplets from scored candidates, runs SkillExtractor,
 * and groups results into topic summaries by tag.
 */

import type { ScoredCandidate } from "./deep";

// ── REMResult ────────────────────────────────────────────────────────────

export interface REMResult {
  /** List of extracted (s, p, o) dicts. */
  triplets: Triplet[];
  /** List of extracted Skill objects. */
  skills: unknown[];
  /** Dict mapping topic tag to summary string. */
  topicSummaries: Record<string, string>;
  /** Total number of triplets extracted. */
  extractedCount: number;
  /** Total number of skills extracted. */
  skillsCount: number;
}

export interface Triplet {
  s: string;
  p: string;
  o: string;
}

export function createREMResult(overrides?: Partial<REMResult>): REMResult {
  return {
    triplets: [],
    skills: [],
    topicSummaries: {},
    extractedCount: 0,
    skillsCount: 0,
    ...overrides,
  };
}

export function remResultToDict(r: REMResult): Record<string, unknown> {
  return {
    extracted_count: r.extractedCount,
    skills_count: r.skillsCount,
    topic_summaries: r.topicSummaries,
    triplets: r.triplets,
  };
}

// ── PatternExtractor ─────────────────────────────────────────────────────

export class PatternExtractor {
  private _llm: unknown;

  constructor(llmProvider?: unknown) {
    this._llm = llmProvider;
  }

  /**
   * Run pattern extraction on scored candidates.
   *
   * @param candidates - Filtered ScoredCandidate list from deep phase.
   * @returns REMResult with triplets, skills, and topic summaries.
   */
  extract(candidates: ScoredCandidate[]): REMResult {
    const result = createREMResult();

    // 1. Build triplets from candidate content
    const triplets = this._buildTriplets(candidates);
    result.triplets = triplets;
    result.extractedCount = triplets.length;

    // 2. Run SkillExtractor (stub -- no dependency in TS)
    //    In the Python version this delegates to SkillExtractor(llm, mode="rule").
    //    Here we produce a simplified stub.
    const skills = this._extractSkills(triplets);
    result.skills = skills;
    result.skillsCount = skills.length;

    // 3. Build topic summaries by tag
    result.topicSummaries = this._buildTopicSummaries(candidates);

    return result;
  }

  // ── triplet construction ───────────────────────────────────────────────

  /**
   * Build simple (subject, predicate, object) triplets from candidates.
   *
   * Uses a simple heuristic: content before a colon is the subject,
   * after the colon is the object. Falls back to word splitting.
   */
  private _buildTriplets(candidates: ScoredCandidate[]): Triplet[] {
    const triplets: Triplet[] = [];
    for (const c of candidates) {
      const content = c.signal.content;
      if (!content) continue;

      let subject: string;
      let obj: string;

      // Simple split heuristic
      const colonIdx = content.indexOf(":");
      const cnColonIdx = content.indexOf("\uff1a");
      const splitIdx = colonIdx >= 0 ? colonIdx : cnColonIdx;

      if (splitIdx >= 0) {
        subject = content.slice(0, splitIdx).trim().slice(0, 50);
        obj = content.slice(splitIdx + 1).trim().slice(0, 100);
      } else {
        const words = content.split(/\s+/);
        if (words.length >= 3) {
          subject = words.slice(0, 2).join(" ").slice(0, 50);
          obj = words.slice(2).join(" ").slice(0, 100);
        } else {
          subject = content.slice(0, 50);
          obj = "";
        }
      }

      triplets.push({ s: subject, p: "relates_to", o: obj });
    }
    return triplets;
  }

  /**
   * Run SkillExtractor on the constructed triplets.
   *
   * Stub implementation -- returns empty list since the Python
   * SkillExtractor dependency is not available in TS.
   */
  private _extractSkills(_triplets: Triplet[]): unknown[] {
    // In Python this imports and invokes:
    //   from claw_mem.extraction.openie_extractor import Triplet
    //   from claw_mem.extraction.skill_extractor import SkillExtractor
    //   extractor = SkillExtractor(llm_provider=self._llm, mode="rule")
    //   return extractor.extract(triplet_objs)
    // For TS we return an empty list as a no-op stub.
    return [];
  }

  // ── topic summaries ────────────────────────────────────────────────────

  /**
   * Group candidates by tags and build concise topic summaries.
   */
  private _buildTopicSummaries(
    candidates: ScoredCandidate[],
  ): Record<string, string> {
    const topicGroups = new Map<string, string[]>();

    for (const c of candidates) {
      const tags = c.signal.tags?.length ? c.signal.tags : ["general"];
      for (const tag of tags) {
        if (!topicGroups.has(tag)) {
          topicGroups.set(tag, []);
        }
        topicGroups.get(tag)!.push(c.signal.content);
      }
    }

    const summaries: Record<string, string> = {};
    for (const [tag, contents] of topicGroups) {
      if (contents.length === 1) {
        summaries[tag] = contents[0].slice(0, 100);
      } else {
        const preview = contents
          .slice(0, 5)
          .map((c) => c.slice(0, 60))
          .join(" | ");
        const suffix =
          contents.length > 5
            ? ` ... (+${contents.length - 5} more)`
            : "";
        summaries[tag] = `[${contents.length} items] ${preview}${suffix}`;
      }
    }

    return summaries;
  }
}
